apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -409,3 +410,183 @@ export const publishResource = async (
+        ),
+    )
+    .selectFrom("resourcePath as rp")
+    .select(["rp.id", "rp.title", "rp.fullPermalink"])
+    .where("rp.id", "=", resourceId)
+    .executeTakeFirst()
+
+  return result
+}
+
+const getResourcesWithLastUpdatedAt = ({ siteId }: { siteId: number }) => {
+  return db
+    .selectFrom("Resource")
+    .select([
+      "Resource.id",
+      "Resource.title",
+      "Resource.type",
+      "Resource.parentId",
+      // To handle cases where either the resource or the blob is updated
+      sql`GREATEST("Resource"."updatedAt", "Blob"."updatedAt")`.as(
+        "lastUpdatedAt",
+      ),
+    ])
+    .leftJoin("Blob", "Resource.draftBlobId", "Blob.id")
+    .where("Resource.siteId", "=", siteId)
+}
+
+const getResourcesWithFullPermalink = async ({
+  resources,
+}: {
+  resources: Omit<SearchResultResource, "fullPermalink">[]
+}): Promise<SearchResultResource[]> => {
+  return await Promise.all(
+    resources.map(async (resource) => ({
+      ...resource,
+      fullPermalink: await getWithFullPermalink({
+        resourceId: resource.id,
+      }).then((r) => r?.fullPermalink ?? ""),
+    })),
+  )
+}
+
+export const getSearchResults = async ({
+  siteId,
+  query,
+  offset,
+  limit,
+}: {
+  siteId: number
+  query: string
+  offset: number
+  limit: number
+}): Promise<{
+  totalCount: number | null
+  resources: SearchResultResource[]
+}> => {
+  const searchTerms: string[] = Array.from(
+    new Set(query.trim().toLowerCase().split(/\s+/)),
+  )
+
+  const queriedResources = getResourcesWithLastUpdatedAt({
+    siteId: Number(siteId),
+  })
+    .where("Resource.type", "in", [
+      // only show user-viewable resources (excluding root page, folder meta etc.)
+      ResourceType.Page,
+      ResourceType.Folder,
+      ResourceType.Collection,
+      ResourceType.CollectionLink,
+      ResourceType.CollectionPage,
+    ])
+    .where((eb) =>
+      eb.or([
+        ...searchTerms.map((searchTerm) =>
+          // Match if the search term is at the start of the title
+          eb("Resource.title", "ilike", `${searchTerm}%`).or(
+            // Match if the search term is in the middle of the title (after a space)
+            eb("Resource.title", "ilike", `% ${searchTerm}%`),
+          ),
+        ),
+      ]),
+    )
+
+  // Currently ordered by number of words matched
+  // followed by `lastUpdatedAt` if there's a tie-break
+  let orderedResources = queriedResources
+  if (searchTerms.length > 1) {
+    orderedResources = orderedResources.orderBy(
+      sql`(
+        ${sql.join(
+          searchTerms.map(
+            (searchTerm) =>
+              // 1. Match if the search term is at the start of the title
+              // 2. Match if the search term is in the middle of the title (after a space)
+              sql`
+                CASE WHEN "Resource"."title" ILIKE ${searchTerm + "%"} THEN 1 ELSE 0 END +
+                CASE WHEN "Resource"."title" ILIKE ${"% " + searchTerm + "%"} THEN 1 ELSE 0 END
```

Comment:
nit: the interpolation I think can put the `%` outside (easier to read but not sure if it works, please check)

```suggestion
                CASE WHEN "Resource"."title" ILIKE ${searchTerm}% THEN 1 ELSE 0 END +
                CASE WHEN "Resource"."title" ILIKE %${searchTerm}% THEN 1 ELSE 0 END
```

Reply:
for some reason on how kysely construct the query, this will not work. suspect it has something to do with how `sql` is parsed in kysely 🥲